### ISSUE TITLE
[CVE-2021-28092] vulnerable version of "is-svg" in dependency chain

### ISSUE THREAD

**Issue by arderyp** [symfony/webpack-encore#954]
The dependency chain relies on is-svg 3.0.0.  I am trying to unwind it, but can't figure out exactly which dependency needs to be updated.  Here's the chain:

```
 @symfony/webpack-encore [1.1.2]
	|--> css-minimizer-webpack-plugin [^1.1.5]
		|--> cssnano [^4.1.10]
			|--> cssnano-preset-default [^4.0.7]
				|--> postcss-svgo [^4.0.2]
					|--> is-svg [^3.0.0]   ***VULNERABLE***
```

[CVE-2021-28092](https://nvd.nist.gov/vuln/detail/CVE-2021-28092) requires `is-svg` to be updated to `4.2.2` or later.

**Comment by arderyp** [comment]
I'm confused.  It looks like `css-minimizer-webpack-plugin` requires `cssnano` version `^4.1.10`.  Yarn is pulling this version for me from https://registry.yarnpkg.com/cssnano/-/cssnano-4.1.10.tgz#0ac41f0b13d13d465487e111b778d42da631b8b2

This version shows:

```
{
  "name": "cssnano",
  "version": "4.1.10",
  ...
  "dependencies": {
    ...
    "cssnano-preset-default": "^4.0.7",
    ...
  },
  ...
  "homepage": "https://github.com/cssnano/cssnano",
  ...
}
```

But the the [NPM package](https://www.npmjs.com/package/cssnano) points to [cssnano/cssnano on github](https://github.com/cssnano/cssnano).

The latest version on the github repo is also `4.1.10`, but it is different, and doesn't rely on `cssnano-preset-default`, which is what ultimately leads to `is-svg`. In fact, it doesn't even have a `dependencies` object in `package.json` but only `devDependencies`.  Also note the name `cssnano-core` vs `cssnano` above:

```
{
  "name": "cssnano-core",
  ...  
  "devDependencies": {
    "@babel/cli": "^7.13.0",
    "@babel/core": "^7.13.8",
    "@babel/node": "^7.13.0",
    "@babel/preset-env": "^7.13.9",
    "all-contributors-cli": "^5.11.0",
    "babel-eslint": "^10.0.1",
    "babel-jest": "^26.6.3",
    "babel-plugin-add-module-exports": "^1.0.2",
    "babel-plugin-module-resolver": "^4.1.0",
    "browserslist": "^4.16.0",
    "camelcase": "^5.3.1",
    "chalk": "^4.1.0",
    "cross-env": "^7.0.3",
    "css-color-names": "^1.0.1",
    "del-cli": "^3.0.1",
    "diff": "^5.0.0",
    "dox": "^0.9.0",
    "eslint": "^7.16.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-prettier": "^3.3.0",
    "eslint-plugin-react": "^7.21.4",
    "fs-extra": "^8.0.1",
    "get-pkg-repo": "^4.1.0",
    "gh-pages": "^1.2.0",
    "glob": "^7.1.4",
    "got": "^11.8.1",
    "html2plaintext": "^2.1.2",
    "jest": "^26.6.3",
    "jest-junit": "^12.0.0",
    "lerna": "4.0.0",
    "mdast-util-heading-range": "^2.1.2",
    "pleeease-filters": "^4.0.0",
    "postcss": "^8.2.1",
    "postcss-devtools": "^1.1.1",
    "postcss-font-magician": "^3.0.0",
    "postcss-scss": "^3.0.4",
    "postcss-simple-vars": "^6.0.1",
    "postcss-value-parser": "^4.1.0",
    "prettier": "^2.2.1",
    "remark": "^10.0.1",
    "remark-behead": "^2.2.1",
    "remark-bookmarks": "^2.1.0",
    "remark-cli": "^6.0.1",
    "remark-frontmatter": "^1.3.1",
    "remark-github": "^7.0.6",
    "remark-heading-gap": "^3.1.0",
    "remark-html": "^9.0.0",
    "remark-license": "^5.0.1",
    "remark-toc": "^5.1.1",
    "stringify-object": "^3.3.0",
    "toml": "^3.0.0",
    "tomlify-j0.4": "^3.0.0",
    "unist-builder": "^1.0.4",
    "unist-util-remove": "^1.0.2",
    "write-file": "^1.0.0"
  },
}
```

**Comment by arderyp** [comment]
see: https://github.com/cssnano/cssnano/issues/1025

**Comment by Kocal** [comment]
Hi,

It looks like [`cssnano@5.0.0-rc2`](https://github.com/cssnano/cssnano/blob/5.0.0-rc2/packages/cssnano/package.json) is already up to date with `cssnano-preset-default` `postcss-svgo` and `is-svg`.

The dependency to fix here is https://github.com/webpack-contrib/css-minimizer-webpack-plugin/, there is already an issue (https://github.com/webpack-contrib/css-minimizer-webpack-plugin/issues/60) and a PR (https://github.com/webpack-contrib/css-minimizer-webpack-plugin/pull/63).

**Comment by arderyp** [comment]
thanks @Kocal !

**Comment by weaverryan** [comment]
Thanks for checking into this @Kocal! There is nothing to do in Encore unless the eventual dependency fix in https://github.com/webpack-contrib/css-minimizer-webpack-plugin/ only happens in new major version of that library (Encore currently requires `^1.1.5`.

**Comment by arderyp** [comment]
@weaverryan, so you think this will resolve this issue, when merged?

https://github.com/webpack-contrib/css-minimizer-webpack-plugin/pull/63

**Comment by weaverryan** [comment]
Yes, that is my impression. Btw, if you read the underlying issue - https://nvd.nist.gov/vuln/detail/CVE-2021-28092 - I'm fairly sure there is no attack vector in Webpack and Encore anyways, which might be why things are moving slowly:

> The is-svg package 2.1.0 through 4.2.1 for Node.js uses a regular expression that is vulnerable to Regular Expression Denial of Service (ReDoS). If an attacker provides a malicious string, is-svg will get stuck processing the input for a very long time.

I can't think of a way for an attacker to supply any string, as all the input is controller by the developer.

**Comment by manooweb** [comment]
> Hi,
> 
> It looks like [`cssnano@5.0.0-rc2`](https://github.com/cssnano/cssnano/blob/5.0.0-rc2/packages/cssnano/package.json) is already up to date with `cssnano-preset-default` `postcss-svgo` and `is-svg`.
> 
> The dependency to fix here is https://github.com/webpack-contrib/css-minimizer-webpack-plugin/, there is already an issue ([webpack-contrib/css-minimizer-webpack-plugin#60](https://github.com/webpack-contrib/css-minimizer-webpack-plugin/issues/60)) and a PR ([webpack-contrib/css-minimizer-webpack-plugin#63](https://github.com/webpack-contrib/css-minimizer-webpack-plugin/pull/63)).

cssnano@5.0.0-rc.2 isn't really ready if I refer to https://github.com/cssnano/cssnano/blob/5.0.0-rc2/packages/postcss-svgo/package.json#L33 
is-svg@4.2.2 is the right patched version.

**Comment by manooweb** [comment]
> Yes, that is my impression. Btw, if you read the underlying issue - https://nvd.nist.gov/vuln/detail/CVE-2021-28092 - I'm fairly sure there is no attack vector in Webpack and Encore anyways, which might be why things are moving slowly:
> 
> > The is-svg package 2.1.0 through 4.2.1 for Node.js uses a regular expression that is vulnerable to Regular Expression Denial of Service (ReDoS). If an attacker provides a malicious string, is-svg will get stuck processing the input for a very long time.
> 
> I can't think of a way for an attacker to supply any string, as all the input is controller by the developer.

I confirm by reading the is-svg@4.2.2 release comment https://github.com/sindresorhus/is-svg/releases/tag/v4.2.2
There is no impact in a Webpack tool context.

**Comment by stof** [comment]
@manooweb a `^4.2.0` constraint still allows installing the patched is-svg version. The fact that the range does not forbid all vulnerable versions is not a blocker for avoiding the vulnerability in your own project.

**Comment by manooweb** [comment]
> @manooweb a `^4.2.0` constraint still allows installing the patched is-svg version. The fact that the range does not forbid all vulnerable versions is not a blocker for avoiding the vulnerability in your own project.

Yes! you're right 😊 `npm update` is sufficient to udpate to the latest patched version and `npm install` will check this at the first install.
Here were blocked because we need an update to a major version from 3 to 4.

**Comment by YannRaulet** [comment]
Hello everyone, if I understand correctly for is-svg 4.2.2, because I am in is-svg 3.0.0, an "npm update is-svg" is sufficient? I'm on Symfony Encore. Or is it better to wait for the next updates? Have a good day

**Comment by stof** [comment]
@YannRaulet no, `npm update` is not sufficient yet, because we need css-minimizer-webpack-plugin to update to the new major version of cssnano (which is in rc for now). So npm won't find a way to resolve dependencies without using an affected version.

**Comment by arderyp** [comment]
@stof is correct, as far as I understand

**Comment by stof** [comment]
This is now resolved by the release of postcss-svgo 4.0.3, which removed the `is-svg` dependency entirely.
